recipes/openai_ab.py

```diff
@@ -0,0 +1,365 @@
+from two experimental conditions (in this case, responses from two
+prompts). The annotator is shown the response pair, without knowing which
+condition produced which response. The annotator marks which one is better,
+and the response is recorded.
+
+At the end of annotation, the results are tallied up, so you can see whether
+one condition produces better results than the other. This lets you apply
+a sound methodology to subjective decisions.
+"""
+from collections import Counter
+import os
+import random
+import sys
+import time
+from pathlib import Path
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, TypeVar, cast
+
+import httpx
+import jinja2
+import prodigy
+import prodigy.components.db
+import prodigy.components.preprocess
+import prodigy.util
+import pydantic
+import srsly
+from dotenv import load_dotenv
+from prodigy.util import msg
+
+_ItemT = TypeVar("_ItemT")
+# Set up openai
+load_dotenv()  # take environment variables from .env.
+
+
+class PromptInput(pydantic.BaseModel):
+    id: str
+    prompt_args: Dict[str, Any]
+
+
+class OpenAIPromptAB:
+    display: jinja2.Template
+    prompts: Dict[str, jinja2.Template]
+    inputs: Iterable[PromptInput]
+    batch_size: int
+    verbose: bool
+    randomize: bool
+    openai_api_org: str
+    openai_api_key: str
+    openai_temperature: float
+    openai_max_tokens: int
+    openai_timeout_s: int
+    openai_n: int
+
+    def __init__(
+        self,
+        display: jinja2.Template,
+        prompts: Dict[str, jinja2.Template],
+        inputs: Iterable[PromptInput],
+        *,
+        openai_api_org: str,
+        openai_api_key: str,
+        openai_model: str,
+        batch_size: int = 10,
+        verbose: bool = False,
+        randomize: bool = True,
+        openai_temperature: float = 0,
+        openai_max_tokens: int = 500,
+        openai_timeout_s: int = 50,
+        openai_n: int = 1,
+    ):
+        self.display = display
+        self.inputs = inputs
+        self.prompts = prompts
+        self.model = openai_model
+        self.batch_size = batch_size
+        self.verbose = verbose
+        self.openai_api_org = openai_api_org
+        self.openai_api_key = openai_api_key
+        self.openai_temperature = openai_temperature
+        self.openai_max_tokens = openai_max_tokens
+        self.openai_timeout_s = openai_timeout_s
+        self.openai_n = openai_n
+        self.randomize = randomize
+
+    def __iter__(self) -> Iterable[Dict]:
+        for input_batch in _batch_sequence(self.inputs, self.batch_size):
+            response_batch = self._get_response_batch(input_batch)
+            for input_, responses in zip(input_batch, response_batch):
+                yield self._make_example(
+                    input_.id,
+                    self.display.render(**input_.prompt_args),
+                    responses,
+                    randomize=self.randomize,
+                )
+
+    def progress(self) -> float:
+        return 0.0
```

Comment:
What does this do?

Reply:
Hey yeah, it's not used anywhere. Removing for now.